src/com/amazon/corretto/crypto/provider/AesGcmSpi.java

```diff
@@ -739,6 +643,153 @@ protected synchronized int engineDoFinal(byte[] bytes, int offset, int length, b
+
+                            // The cost of calling decryptAADBuf.getDataBuffer() when its buffer is empty is significant for 16-byte
+                            // decrypt operations (approximately a 7% performance hit). To avoid this, we reuse the same empty array
+                            // instead in this common-case path.
+                            decryptAADBuf.size() != 0 ? decryptAADBuf.getDataBuffer() : EMPTY_ARRAY,
+                            decryptAADBuf.size()
+                    );
+                });
+            } else {
+                // We don't have an existing context, however we might want to save one
+                final long[] ptrOut = keyUsageCount > KEY_REUSE_THRESHOLD ? new long[1] : null;
+                outLen = oneShotDecrypt(
+                        0,
+                        ptrOut,
+                        workingInputArray,
+                        workingInputOffset,
+                        workingInputLength,
+
+                        result,
+                        0,
+
+                        tagLength,
+                        key,
+                        iv,
+
+                        // The cost of calling decryptAADBuf.getDataBuffer() when its buffer is empty is significant for 16-byte
+                        // decrypt operations (approximately a 7% performance hit). To avoid this, we reuse the same empty array
+                        // instead in this common-case path.
+                        decryptAADBuf.size() != 0 ? decryptAADBuf.getDataBuffer() : EMPTY_ARRAY,
+                        decryptAADBuf.size()
+                );
+
+                if (ptrOut != null) {
+                    context = new NativeContext(ptrOut[0]);
+                }
+            }
+            // Decryption completed successfully.
+            if (outLen == result.length) {
+                // This should always be the case.
+                return result;
+            } else if (outLen < result.length) {
+                // While technically allowed by the APIs, this should never happen.
+                return Arrays.copyOf(result, outLen);
+            } else {
+                throw new RuntimeCryptoException("Output longer than expected.");
+            }
+        } finally {
+            stateReset();
+        }
+    }
+
+    @Override
+    protected byte[] engineDoFinal(byte[] bytes, int offset, int length)
+            throws IllegalBlockSizeException, BadPaddingException {
+        if (opMode == NATIVE_MODE_ENCRYPT) {
+            final byte[] buf = new byte[engineGetOutputSize(length)];
+            int actualLength;
+            try {
+                actualLength = engineEncryptFinal(bytes, offset, length, buf, 0);
+            } catch (ShortBufferException e) {
+                throw new AssertionError(e);
+            }
+
+            if (actualLength == buf.length) {
+                return buf;
+            } else {
+                return Arrays.copyOf(buf, actualLength);
```

Comment:
what would cause `engineGetOutputSize` to be incorrect in the encryption case? isn't it just len(pt) + padding + tagLen?

Reply:
It should never happen but libcrypto is technically allowed to do this. A prior version had a comment describing this. I'll re-add it.